File: midi-b/midi_b/mapping.py
"""Mapping CSV export for MIDI-B submission artifacts."""

from __future__ import annotations

import csv
import io


class MappingCollector:
    """Accumulates UID and patient ID mappings across multiple files."""

    def __init__(self) -> None:
        self.uid_mappings: dict[str, str] = {}
        self.patient_id_mappings: dict[str, dict[str, str | int | None]] = {}

    def add_uid_mappings(self, mappings: dict[str, str]) -> None:
        """Merge UID mappings from a single file de-identification."""
        self.uid_mappings.update(mappings)

    def add_patient_id_mapping(
        self,
        original: str,
        replacement: str,
        date_shift_offset: int | None = None,
    ) -> None:
        """Record a patient ID → pseudonym mapping."""
        self.patient_id_mappings[original] = {
            "replacement": replacement,
            "date_shift_offset": date_shift_offset,
        }

    def to_uid_csv(self) -> str:
        """Export UID mappings as CSV text."""
        buf = io.StringIO()
        writer = csv.writer(buf)
        writer.writerow(["original_uid", "replacement_uid"])
        for original, replacement in sorted(self.uid_mappings.items()):
            writer.writerow([original, replacement])
        return buf.getvalue()

    def to_patient_id_csv(self) -> str:
        """Export patient ID mappings as CSV text."""
        buf = io.StringIO()
        writer = csv.writer(buf)
        writer.writerow(["original_patient_id", "replacement_patient_id", "date_shift_offset"])
        for original, info in sorted(self.patient_id_mappings.items()):
            writer.writerow([
                original,
                info["replacement"],
                info["date_shift_offset"] if info["date_shift_offset"] is not None else "",
            ])
        return buf.getvalue()

File: client/src/dicom/mapping.ts
/**
 * De-identification mapping collector and CSV exporter.
 *
 * Accumulates UID and PatientID mappings across multiple deidentify() calls
 * and exports them as CSV files matching the MIDI-B submission format.
 */

import type { DeidResult } from './deid'

export interface DeidMappings {
  /** Original UID → replacement UID */
  uids: Map<string, string>
  /** Original PatientID → pseudonym */
  patientIds: Map<string, string>
  /** PatientID → date shift offset in days (if date shifting was applied) */
  dateOffsets: Map<string, number>
}

export interface MappingCollector {
  /** Record mappings from a single deidentify() result */
  record(result: DeidResult): void
  /** Get all accumulated mappings */
  getMappings(): DeidMappings
  /** Export UID mappings as CSV: "original_uid,replacement_uid" */
  toUidCsv(): string
  /** Export PatientID mappings as CSV: "original_patient_id,replacement_patient_id,date_shift_offset" */
  toPatientIdCsv(): string
}

export function createMappingCollector(): MappingCollector {
  const uids = new Map<string, string>()
  const patientIds = new Map<string, string>()
  const dateOffsets = new Map<string, number>()

  return {
    record(result: DeidResult): void {
      // Collect UID mappings
      if (result.uidMappings) {
        for (const [original, replacement] of result.uidMappings) {
          uids.set(original, replacement)
        }
      }

      // Collect PatientID mapping
      if (result.patientIdMapping) {
        patientIds.set(result.patientIdMapping.original, result.patientIdMapping.replacement)
      }

      // Collect date shift offset
      if (result.dateShiftOffset !== undefined && result.patientIdMapping) {
        dateOffsets.set(result.patientIdMapping.original, result.dateShiftOffset)
      }
    },

    getMappings(): DeidMappings {
      return {
        uids: new Map(uids),
        patientIds: new Map(patientIds),
        dateOffsets: new Map(dateOffsets),
      }
    },

    toUidCsv(): string {
      const lines = ['original_uid,replacement_uid']
      for (const [original, replacement] of uids) {
        lines.push(`${original},${replacement}`)
      }
      return lines.join('\n')
    },

    toPatientIdCsv(): string {
      const lines = ['original_patient_id,replacement_patient_id,date_shift_offset']
      for (const [original, replacement] of patientIds) {
        const offset = dateOffsets.get(original) ?? ''
        lines.push(`${original},${replacement},${offset}`)
      }
      return lines.join('\n')
    },
  }
}
